mprotect_builder.cc

```diff
@@ -1,4 +1,70 @@
 #include "mprotect_builder.h"
 
-constexpr uint8_t MprotectBuilder::memprotect_body_code[];
-constexpr uint8_t MprotectBuilder::memprotect_end_code[];
+constexpr uint8_t MprotectBuilder::memprotect_body_code_x86_64[];
+constexpr uint8_t MprotectBuilder::memprotect_end_code_x86_64[];
+constexpr uint8_t MprotectBuilder::memprotect_end_code_aarch64[];
+
+void MprotectBuilder::EmitX86_64(FILE* fp, uintptr_t mprotect_code_offset) {
+    uint8_t* mprotect_code = (uint8_t*)malloc(Size());
+    uint8_t* mprotect_code_head = mprotect_code;
+    for (int i = 0; i < offsets.size(); i++) {
+        LOG(INFO) << "MprotectBuilder::Emit: " << SOLD_LOG_BITS(offsets[i]) << SOLD_LOG_BITS(sizes[i]);
+
+        memcpy(mprotect_code_head, memprotect_body_code_x86_64, sizeof(memprotect_body_code_x86_64));
+        int64_t* offset_p = (int64_t*)(mprotect_code_head + memprotect_body_addr_offset_x86_64);
+        // 17 is the length of `mov $0xdeadbeefdeadbeef, %rdi; lea (%rip), %rsi'
+        int64_t offset_v = (offsets[i] & (~(0x1000 - 1))) -
+                           (static_cast<int64_t>(mprotect_code_offset) + static_cast<int64_t>(mprotect_code_head - mprotect_code) + 17);
+        LOG(INFO) << "MprotectBuilder::Emit" << SOLD_LOG_BITS(offset_v) << SOLD_LOG_KEY(offset_v);
+        *offset_p = offset_v;
+        uint32_t* size_p = (uint32_t*)(mprotect_code_head + memprotect_body_size_offset_x86_64);
+        *size_p = ((offsets[i] + sizes[i]) & (~(0x1000 - 1))) - (offsets[i] & (~(0x1000 - 1)));
+        mprotect_code_head += sizeof(memprotect_body_code_x86_64);
+    }
+    memcpy(mprotect_code_head, memprotect_end_code_x86_64, sizeof(memprotect_end_code_x86_64));
+    mprotect_code_head += sizeof(memprotect_end_code_x86_64);
+
+    CHECK(fwrite(mprotect_code, sizeof(uint8_t), Size(), fp) == Size());
+    free(mprotect_code);
+}
+
+namespace {
+std::vector<uint32_t> set_immediate_to_register_aarch64(int64_t value, uint8_t reg) {
+    std::vector<uint32_t> ret;
+    for (int i = 0; i < 4; i++) {
+        uint32_t u = 0b11110010'10000000'00000000'00000000;  // movk instruction
+        u |= (i << 21);
+        u |= ((value >> (i * 16)) & 0xffff) << 5;
+        u |= reg;
+        ret.emplace_back(u);
+    }
+    return ret;
+}
+}  // namespace
+
+void MprotectBuilder::EmitAarch64(FILE* fp, uintptr_t mprotect_code_offset) {
```

Comment:
[optional] I think we are assuming both host and target are little endian. Perhaps it's worth adding a few checks? It's not important at all given everyone is using little endian, though.

Reply:
It sounds good idea! https://github.com/shinh/sold/pull/80/commits/87f45ac2bc0eaa57cc11943ffdc0170f73b1dff9